tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/package.scala

```diff
@@ -1,22 +1,217 @@
+ * distinct layers:
+ *   1. The signature layer
+ *   1. The type-safe, scope-unsafe layer
+ *   1. The type-safe, scope-safe layer
  *
+ * ==The signature layer==
+ *
+ * Each TLA+ operator has an associated [[TlaOper]] operator in the IR. The majority (but not all) of operators have
+ * type signatures, that is, they restrict the types of the arguments, that may be used to construct valid [[OperEx]]
+ * expressions. For instance, [[TlaArithOper.plus]] represents the arithmetic operator `+` and has the signature `(Int,
+ * Int) => Int` in the type system, meaning that `e @ OperEx(TlaArithOper.plus, x, y)` is considered valid iff `x`, `y`,
+ * and `e` are all tagged with `Typed(IntT1)`. Similarly, [[TlaOper.chooseBounded]] has the signature `(t, Set(t), Bool)
+ * \=> t`, meaning that `e @ OperEx(TlaOper.chooseBounded, x, S, p)` is considered valid if `x` and `e` are tagged with
+ * `Typed(t)`, for some `t`, `S` is tagged with `Typed(SetT1(t))` (for the same `t`), and `p` is tagged with
+ * `Typed(BoolT1)`.
+ *
+ * You can think of a signature `(a1,...,an) => b` as a partial function; given a sequence of argument types `v1, ...
+ * vn`, it returns some type `c` (derived from `b`), if the types `v1, ..., vn` match the pattern `a1, ..., an`. More
+ * precisely, the return value is `s(b)`, if there exists some type substitution `s`, such that `s(ai) = vi` for all `i
+ * \= 1,...,n`.
+ *
+ * To reason about signatures, we define two types, [[PartialSignature]] and [[Signature]]. They serve the same purpose,
+ * but differ in the way they behave on mis-matches. [[PartialSignature]]s are bare-bones descriptions of the
+ * happy-case; given an input pattern `a1,...,an`, they describe the output `b`. For example, `plus` has the
+ * [[PartialSignature]]
+ * {{{
+ *   { case Seq(IntT1, IntT1) => IntT1 }
+ * }}}
+ * that is, given an input sequence of exactly two arguments, both of which are `IntT1`, it deduces the result type to
+ * be `IntT1`. Similarly, `chooseBounded` has the signature
+ * {{{
+ *   { case Seq(t, SetT1(tt), BoolT1) if t == tt => t }
+ * }}}
+ * that is, given an input sequence of exactly three arguments, the first two of which are `t` and `Set(t)`, for some
+ * `t`, and the last of which is `BoolT1`, it deduces the result type to be `t`. This means that applying this partial
+ * signature to `Seq(IntT1, SetT1(IntT1), BoolT1)`, would return `IntT1` while applying it to `Seq(StrT1, SetT1(StrT1),
+ * BoolT1)` would return `StrT1`. It is not applicable to `Seq(IntT1, SetT1(StrT1), BoolT1)`.
+ *
+ * [[Signature]]s complete [[PartialSignature]]s, by equipping them with an exception message, to be thrown should the
+ * input sequence not match the pattern defined by the [[PartialSignature]]. This lifting is done by
+ * [[BuilderUtil.completePartial completePartial]], if all we want is a total function, or
+ * [[BuilderUtil.checkForArityException checkForArityException]], to output a more precise error message, outlining
```

Comment:
When would I ever want to interact with `completePartial`? Does it make sense to expose it publicly? The paragraph below sounds as if even internally, `checkForArityException` is used.

Reply:
Well, if you're working underneath a `require(arityConstraint)` block you wouldn't need to also check for arity. But you're right, in principle we could hide that one and just expose the arity method.